files/unarchive.py

```diff
@@ -539,23 +536,24 @@ def __init__(self, src, dest, file_args, module):
         self.compress_mode = 'gz'
         self._files_in_archive = []
 
-    def _get_tar_fileobj(self):
-        """Returns a file object that can be read by ``tarfile.open()``."""
-        return open(self.src, 'rb')
-
     @property
     def files_in_archive(self, force_refresh=False):
         if self._files_in_archive and not force_refresh:
             return self._files_in_archive
 
-        # The use of Python's tarfile module here allows us to easily avoid tricky file encoding
-        # problems. Ref #11348
-        try:
-            tf = tarfile.open(fileobj=self._get_tar_fileobj(), mode='r:%s' % self.compress_mode)
-        except Exception:
+        cmd = '%s -t%s' % (self.cmd_path, self.zipflag)
+        if self.opts:
+            cmd += ' ' + ' '.join(self.opts)
+        if self.excludes:
+            cmd += ' --exclude="' + '" --exclude="'.join(self.excludes) + '"'
+        cmd += ' -f "%s"' % self.src
+        rc, out, err = self.module.run_command(cmd)
+        if rc != 0:
             raise UnarchiveError('Unable to list files in the archive')
 
-        for filename in tf.getnames():
+        for filename in out.splitlines():
+            # Compensate for locale-related problems in gtar output (octal unicode representation) #11348
+            filename = filename.decode('string_escape')
```

Comment:
We can't use string_escape as there's no way to make it compatible with python3.  There may also be encoding issues as string_escape always uses latin1 as its encoding even though the string will usually be utf-8 but I'm not certain about that.  I started looking into this and finally found the "unicode_escape doesn't work in general" answer to http://stackoverflow.com/questions/4020539/process-escape-sequences-in-a-string-in-python  We had made a variation on that work in the controller code.  I struggled to adapt it to this situation for a bit before #3575 was submitted.  That seemed like an easier to understand solution than fighting with a finicky regex so I abandoned that effort when we merged the tarfile solution.